chalk-solve/src/logging_db.rs

```diff
@@ -327,45 +325,45 @@ where
     P: Borrow<DB> + Debug,
 {
     fn custom_clauses(&self) -> Vec<chalk_ir::ProgramClause<I>> {
-        self.db.borrow().custom_clauses()
+        self.db.custom_clauses()
     }
 
     fn associated_ty_data(
         &self,
         ty: chalk_ir::AssocTypeId<I>,
     ) -> Arc<crate::rust_ir::AssociatedTyDatum<I>> {
-        self.db.borrow().associated_ty_data(ty)
+        self.db.associated_ty_data(ty)
     }
 
     fn trait_datum(&self, trait_id: TraitId<I>) -> Arc<TraitDatum<I>> {
-        self.db.borrow().trait_datum(trait_id)
+        self.db.trait_datum(trait_id)
     }
 
     fn adt_datum(&self, adt_id: AdtId<I>) -> Arc<AdtDatum<I>> {
-        self.db.borrow().adt_datum(adt_id)
+        self.db.adt_datum(adt_id)
     }
 
     fn adt_repr(&self, id: AdtId<I>) -> AdtRepr {
-        self.db.borrow().adt_repr(id)
+        self.db.adt_repr(id)
     }
 
     fn impl_datum(&self, impl_id: ImplId<I>) -> Arc<ImplDatum<I>> {
-        self.db.borrow().impl_datum(impl_id)
+        self.db.impl_datum(impl_id)
     }
 
     fn associated_ty_value(
         &self,
         id: crate::rust_ir::AssociatedTyValueId<I>,
     ) -> Arc<crate::rust_ir::AssociatedTyValue<I>> {
-        self.db.borrow().associated_ty_value(id)
+        self.db.associated_ty_value(id)
     }
 
     fn opaque_ty_data(&self, id: OpaqueTyId<I>) -> Arc<OpaqueTyDatum<I>> {
-        self.db.borrow().opaque_ty_data(id)
+        self.db.opaque_ty_data(id)
     }
 
     fn hidden_opaque_type(&self, id: OpaqueTyId<I>) -> Ty<I> {
-        self.db.borrow().hidden_opaque_type(id)
+        self.db.hidden_opaque_type(id)
     }
 
     fn impls_for_trait(
```

Comment:
You missed a `db.borrow()` call here, and in `closure_inputs_and_output` as well